### DISCUSSION TITLE
Add multiple environment variables

### DISCUSSION THREAD

**Discussion by mosesoak** [vercel/vercel#4977] [Ideas] [answered]
I have a bundle of about 15 env vars I've set up locally and now want to quickly port them to the admin panel using `vercel env add`.  However, adding them via the cli does not provide any way to automate this. 

Adding from a file doesn't help (it uses the whole file), and adding individually is as manual of a copy/paste process as from the panel (somewhat more actually since I have to type the environment name each time).

What's needed is a way to publish multiple vars from a single file.

Thanks for considering the feature add!

**Comment by styfle** [comment, marked as answer]
We have a feature request to add a `vc env push` command to complement the `vc env pull` command. This is tracked internally as Story 2195.

However, in the meantime, you can write a bash script to loop over your env file key/values and call `vc env add` which accepts all the parameters you need.

```sh
echo 'env_value_here' | vc env add ENV_NAME_HERE production
```

This is how @jeantil did it for production env:

```sh
for key in $(echo $keys);do                                                                                                                                                           
    vc env rm ${key} production
    echo -n "${(P)key}" | vc env add $key production
done
```

Try `vc env --help` for more info.

**Reply by sovattha** [reply]
@styfle When I execute 

```
echo 'whatevervalue' | vc env add MY_ENV_VARIABLE development
```

When I `vc env pull test.txt`, I get a trailing `\n` in my test.txt file :

```
# Created by Vercel CLI
MY_ENV_VARIABLE="whatevervalue\n"
```

I work on Mac OS and my vc version is Vercel CLI 20.1.2

**Reply by styfle** [reply]
Thanks, looks like a bug. We'll get this fixed!

In the meantime, you can use `vc env add` without an additional parameters and it works correctly.

**Reply by adharshmk96** [reply]
how can we add a particular env to multiple environments ( production and preview )

**Reply by styfle** [reply]
You'll have to use the normal prompts for `vc add` which has checkboxes for multiple environments or you can use the stdin method and add it multiple times, once for each environment.

**Reply by jasongoetz** [reply]
The issue reported by @sovattha is still an issue. Newline characters still show up with the use of echo, which is one of the recommended usages recommended here: https://vercel.com/docs/cli/env#extended-usage.

It can be fixed easily enough by adding `| tr -d '\n'` after the echo command, but it still can take some time to troubleshoot that one as I found out.

**Comment by Pytal** [comment]
@mosesoak I encountered this issue as well and wanted a nice CLI to automate this so I made [deploy-env](https://github.com/Pytal/deploy-env), feel free to try it out and I hope it suits your needs until we get `vc env push`!

**Comment by agcty** [comment]
When will vc env push be available?

**Comment by maddyonline** [comment]
Here is a bash script I created to push local dev env variables to Vercel env.
```bash
REACT_APP_FIREBASE_API_KEY="a"
REACT_APP_FIREBASE_AUTH_DOMAIN="b"
REACT_APP_FIREBASE_DATABASE_URL="c"
REACT_APP_FIREBASE_PROJECT_ID="d"
REACT_APP_FIREBASE_STORAGE_BUCKET="e"
REACT_APP_FIREBASE_MESSAGING_SENDER_ID="f"
REACT_APP_FIREBASE_APP_ID="g"
REACT_APP_FIREBASE_MEASUREMENT_ID="h"

for var in  \
REACT_APP_FIREBASE_API_KEY \
REACT_APP_FIREBASE_AUTH_DOMAIN \
REACT_APP_FIREBASE_DATABASE_URL \
REACT_APP_FIREBASE_PROJECT_ID \
REACT_APP_FIREBASE_STORAGE_BUCKET \
REACT_APP_FIREBASE_MESSAGING_SENDER_ID \
REACT_APP_FIREBASE_APP_ID \
REACT_APP_FIREBASE_MEASUREMENT_ID \
  ; do
  echo $var="${!var}"
  echo "vercel env rm -y $var development >/dev/null 2>&1 || true && echo -n ${!var} | vercel env add $var development"
  vercel env rm -y $var development >/dev/null 2>&1 || true && echo -n ${!var} | vercel env add $var development
done
```

**Comment by RoyalSix** [comment]
.env.production
```env
FOO=bar
```


vercel-env-push.sh
```sh
set -a
. $1
set +a
env_vars=`cat "$1" | sed s/=.*//g`
for key in $(echo $env_vars);do 
    sleep 5 #For vercel API
    vc env rm ${key} $2  -y                                                                                                                         
    echo "${!key}" | vc env add plain $key $2
done
exit 0;
```

package.json
```json
  "scripts": {
     //...your scripts
    "vercel:env:push": "bash scripts/env.sh"
  },
```

usage:
1st param path
2nd param environment type
`yarn vercel:env:push .env.production production`

<hr />

Note: use with caution

**Comment by GorvGoyl** [comment]
Would really appreciate some update on this. we really need a seamless way to work on different environments.

**Comment by iShubhamPrakash** [comment]
I have updated @RoyalSix script to skip for the commented values-

```bash
set -a
. $1
set +a
env_vars=`cat "$1" | sed s/=.*//g`

for key in $(echo $env_vars);do

if [[ $key =~ ^#.* ]];then
    echo "Skipping the commented value" - $key

else
    echo "Uploading" - $key
    #sleep 2 #For vercel API
    vc env rm ${key} $2  -y
    echo "${!key}" | vc env add $key $2
fi
done

exit 0;
```

### Caveats
- Don't use spaces between = and values  

e.g
```

NAME="value"  #✅
NAME = "value"  #❌

```

**Comment by ryan-blunden** [comment]
You can easily do this in a [single line of code](https://blog.doppler.com/how-to-add-multiple-environment-variables-to-vercel) by using environment variables in JSON format combined with `jq` and the Vercel CLI.

**Comment by adikari** [comment]
the solution with `env add` works fine. however, this is not ideal as it takes a bit of time to add each environment variable one at a time. It would be good to have a command that can push multiple keys in 1 request. something like:

```
vercel env push multiplevars.json
```

**Comment by adoprog** [comment]
We are creating multiple projects (each has 15+ env. variables) and hitting the API limits (429 returned) because of this. Is there any way to push multiple environment variables in single API call?

**Reply by jeanfrancoislarente** [reply]
@styfle are you able to comment on this use case or tag someone who might be able to? Bulk adding env variables would solve our problem. Our alternative is to throttle our deployments which would exponentially increase our deployment times.

**Reply by iShubhamPrakash** [reply]
For now, I guess this is possible only through 3rd party integrations like Doppler.

**Reply by styfle** [reply]
Please contact support so our team can better understand your particular use case https://vercel.com/support/request

Thanks!

**Comment by HiDeoo** [comment]
I also encountered the same issue, specially the need to push environment variables from a .env file to **multiple environments** so I wrote [my own tool to do that either through the CLI or an API](https://github.com/HiDeoo/vercel-env-push) but indeed, an official command to push multiple environment variables to multiple environments in a single request would be the best way to do it.

**Comment by t3dotgg** [comment]
This just embarrassed and frustrated the shit out of me on stream. Please, like, do _something_

**Reply by maddyonline** [reply]
https://www.doppler.com/ has a nice Vercel integration which automatically syncs env variables there. Do check it out!

**Reply by leerob** [reply]
We are working on this 🙏 Realized I haven't posted anything back here.

**Reply by JayArrowz** [reply]
> This just embarrassed and frustrated the shit out of me on stream. Please, like, do _something_

Same annoyed, how did anyone think its logical to add env vars one at a time. Even the UI doesn't let you bulk import a JSON or .env format. Sitting here spending several hours adding env vars for multiple environments

**Reply by adoprog** [reply]
If you need it when creating a new project, the API call actually does accept "environmentVariables" property which is a list that can contain any number of variables. Haven't seen it documented but this is how their own UI works too.

**Reply by amyegan** [reply]
Here's a link to that API doc: [Create one or more environment variables](https://vercel.com/docs/rest-api#endpoints/projects/create-one-or-more-environment-variables)

**Reply by adoprog** [reply]
@amyegan It says "or more" but does not provide sample payload for multiple.

**Reply by amyegan** [reply]
Hi @adoprog. The request body should be a JSON array, so you can add multiple items in the same call. The body could look something like 
```
[
  {
    "key": "TEMP_VAR1",
    "target": ["development"],
    "type": "plain",
    "value": "abcde"
  },
  {
    "key": "TEMP_VAR2",
    "target": ["preview", "development"],
    "type": "plain",
    "value": "lmnopq"
  }
]
```

**Reply by adoprog** [reply]
Thanks!

**Comment by BrennerSpear** [comment]
Bumping here that this would be nice. Even just being able to copy and paste a .env into the UI or something

**Comment by Gilighost** [comment]
Would really love to see support for this because managing many environment variables through the UI can be quite tedious.
Right now I am using this command with the (won't work if you have comments in your .env but could be modified to allow that): 
`cat .env | sed 's/=/ /' | xargs -n 2 bash -c 'echo -n $1 | vercel env add $0'`

It would also be great if there was support for updating existing environment variables with the CLI without having to remove them first.

**Comment by jgtvares** [comment]
Hey everyone!
My use case is that on every push to `main` or `develop` branches I want my environment variables to be synced to Vercel, so I don't need to keep adding them manually every time.
I use **AWS Parameter Store** as the source of truth, so the variables are stored there and loaded to a `.env` file with a script I created.
So, it would be really nice to be able to upload the entire `.env` file to Vercel, just like Doppler supports.
But, since Vercel didn't solve this yet I've decided to put some effort to create something more elegant on my own.
I developed a [**JavaScript Github Action**](https://github.com/marketplace/actions/vercel-envs-push) to upload variables in a `.env` format!
It's working great, but it has some limitations which are described in the README.
Feel free to contribute to the project to make it less limited!

Hope it can help someone!

**Comment by olawalejuwonm** [comment]
The below script was useful

```
# This scripts iterates the .env file pass as first argument and environment as second argument
# and pushes the variables to the Vercel environment
# Usage: ./vercel-env-push.sh .env production

# Check if the .env file exists
if [ ! -f "$1" ]; then
  echo "File $1 does not exist"
  exit 1
fi

# Check if the environment is passed
if [ -z "$2" ]; then
  echo "Environment is not passed"
  exit 1
fi

# Read the .env file
while read -r line; do
  # Check if the line is not empty
  if [ -n "$line" ]; then
    # Check if the line is not a comment
    if [[ ! $line =~ ^#.*$ ]]; then
      # Split the line by the equal sign
      IFS='=' read -ra VAR <<< "$line"
      echo "Setting ${VAR[0]} to ${VAR[1]}"
      #check if ${VAR[0]} is not a comment and if it is not empty
        if [[ ! ${VAR[0]} =~ ^#.*$ ]] && [ -n "${VAR[0]}" ]; then
            # Push the variable to the Vercel environment
            vc env rm ${VAR[0]} $2  -y 
            echo "${VAR[1]}" | vercel env add ${VAR[0]} $2
            echo "vercel env add ${VAR[0]} ${VAR[1]} -e $2"
        fi
    fi
  fi
done < "$1"
```

**Comment by OR13** [comment]
Better examples on https://vercel.com/docs/cli/env would really help.

What is [name] ?

What is [environment] ?

**Reply by amyegan** [reply]
Thanks for pointing this out! I'll open a ticket to add more specific examples to the CLI docs
